### PR TITLE
Add missing builtins

### PR DESCRIPTION
**Pull request by davidar**
I added a bunch of WMA builtins that aren't in Mathics yet:

- NumberDigit
- BellB
- EulerE
- JacobiSymbol
- KroneckerSymbol
- LucasL
- PolygonalNumber
- SquaresR
- LinearRecurrence
- ~~RootSum~~ I think this one needs some more work for converting to sympy
- SeriesCoefficient
- DivisorSigma
- DivisorSum
- IntegerPart
- IntegerPartitions
- MersennePrimeExponent
- MoebiusMu
- PowersRepresentations
- HypergeometricU
- LambertW
- Subfactorial
- PolyLog
- ReverseSort

I also expanded the functionality of a few existing builtins to match the behaviour of WMA:

- handle non-integer bounds in Sum
- fixed a bug in RealDigits where it would report the wrong number of digits for powers of 10 (or whatever base is specified)
- support Fibonacci polynomials
- support symbolic bounds in Table (and other IterationFunctions)
- support providing SeriesData to CoefficientList
- added the listable attribute to a few functions
- allow supplying a precision when calling `N[]` on a complex number

I can split some of these out into separate PRs if they need more work to merge.

I've started setting up some automated tests to pull all the Mathematica programs listed in OEIS and check that Mathics can evaluate them to the correct sequence values - these are all just the low hanging fruit that fell out of doing that for the first couple of hundred sequences.

### CODE REVIEW

**Comment by rocky** [comment]
Wow - thanks. I'll try to look at over the weekend.

File: mathics/builtin/intfns/combinatorial.py [hunk 1]
```diff
@@ -214,6 +273,89 @@ def _compute(self, n, c_ff, c_ft, c_tf, c_tt):
...
+    <dl>
+      <dt>'KroneckerSymbol[$a$, $n$]'
+      <dd>returns the Kronecker symbol $\\left(\\frac{a}{n}\\right)$.
+    </dl>
+
+    >> Table[KroneckerSymbol[n, m], {n, 5}, {m, 5}]
+     = {{1, 1, 1, 1, 1}, {1, 0, -1, 0, -1}, {1, -1, 0, 1, -1}, {1, 0, 1, 0, 1}, {1, -1, -1, 1, 0}}
+    """
+
+    summary_text = "Kronecker symbol"
```
Comment (axkr) [inline mathics/builtin/intfns/combinatorial.py:310]: `KroneckerSymbol` has attribute `Listable`

**Comment by rocky** [comment]
One small addition - Update [CHNAGES.rst](https://github.com/Mathics3/mathics-core/blob/master/CHANGES.rst#new-builtins). You can also add your name to these changes.  See `ComplexExpand[]` for how this was done previously, but if you want to add a name rather than a handle. that's fine too.

File: mathics/builtin/intfns/combinatorial.py [hunk 2]
```diff
@@ -184,6 +214,35 @@ def _compute(self, n, c_ff, c_ft, c_tf, c_tt):
...
+    </dl>
+
+    >> Table[EulerE[k], {k, 0, 10}]
+     = {1, 0, -1, 0, 5, 0, -61, 0, 1385, 0, -50521}
+
+    >> EulerE[5, z]
+     = -1 / 2 + 5 z ^ 2 / 2 - 5 z ^ 4 / 2 + z ^ 5
+    """
+
+    attributes = A_LISTABLE | A_NUMERIC_FUNCTION | A_PROTECTED
```
Comment (axkr) [inline mathics/builtin/intfns/combinatorial.py:237]: `EulerE` doesn't contain attribute `NumericFunction`

**Comment by mmatera** [comment]
@davidar thanks for all this work! I will try to review this over the weekend.

File: mathics/builtin/atomic/numbers.py [hunk 3]
```diff
@@ -60,7 +60,9 @@
 
 @lru_cache()
 def log_n_b(py_n, py_b) -> int:
-    return int(mpmath.ceil(mpmath.log(py_n, py_b))) if py_n != 0 and py_n != 1 else 1
+    return (
```
Comment (mmatera) [inline mathics/builtin/atomic/numbers.py:63]: This could be moved to mathics.eval.numbers

File: mathics/builtin/intfns/combinatorial.py [hunk 4]
```diff
@@ -39,6 +40,35 @@
...
+    </dl>
+
+    >> BellB[10]
+     = 115975
+
+    >> BellB[5, x]
+     = x + 15 x ^ 2 + 25 x ^ 3 + 10 x ^ 4 + x ^ 5
+    """
+
+    attributes = A_LISTABLE | A_NUMERIC_FUNCTION | A_PROTECTED
```
Comment (mmatera) [inline mathics/builtin/intfns/combinatorial.py:63]: In WMA, `Attributes[BellB]={Listable, NHoldFirst, NumericFunction, Protected, ReadProtected}`

Comment (mmatera) [reply mathics/builtin/intfns/combinatorial.py:63]: (`ReadProtected` is not needed, but `NHoldFirst` affects the evaluation)

File: mathics/builtin/intfns/combinatorial.py [hunk 5]
```diff
@@ -352,6 +520,40 @@ def _compute(self, n, c_ff, c_ft, c_tf, c_tt):
...
+     = {5, 13, 29, 49, 81}
+    >> Table[SquaresR[4, n], {n, 10}]
+     = {8, 24, 32, 24, 48, 96, 64, 24, 104, 144}
+    >> Table[SquaresR[6, n], {n, 10}]
+     = {12, 60, 160, 252, 312, 544, 960, 1020, 876, 1560}
+    >> Table[SquaresR[8, n], {n, 10}]
+     = {16, 112, 448, 1136, 2016, 3136, 5504, 9328, 12112, 14112}
+    """
+
+    attributes = A_LISTABLE | A_NUMERIC_FUNCTION | A_PROTECTED
```
Comment (mmatera) [inline mathics/builtin/intfns/combinatorial.py:545]: In WMA, 
`Attributes[SquaresR]={Protected, ReadProtected}`

File: mathics/builtin/intfns/recurrence.py [hunk 6]
```diff
@@ -47,6 +51,11 @@ class Fibonacci(MPMathFunction):
     mpmath_name = "fibonacci"
     summary_text = "Fibonacci's numbers"
 
+    rules = {
```
Comment (mmatera) [inline mathics/builtin/intfns/recurrence.py:54]: ``Attributes[Fibonacci]={Listable, NumericFunction, Protected, ReadProtected}``

**Comment by davidar** [comment]
Good point, I'll go through and fix up all the attributes. Though I'm a bit confused about some of the discrepancies, do these change between different versions or what?

image

File: mathics/builtin/atomic/numbers.py (hunk 3 above)
Comment (rocky) [reply mathics/builtin/atomic/numbers.py:63]: Let me elaborate a little on @mmatera' s comment.   Right now, we have a tree evaluator for Mathics3, and down the line if we are to get better performance, we will want an instruction interpreter. 

Moving evaluation routines somewhere under `mathics.eval`  (and here it would be `mahics.eval.atomic.numbers`) will help out when such large task is done. 

We are moving towards an organization where what is in `mathics.builtin` are the classes that implement the Builtin functions and inside their `eval()` routines we have largely the builtin properties (rules, attributes, doctest examples, function call matching signature, etc.)  and some parameter checking.  The non OO stuff and the bulk of the evaluation then is moved into `mathics.eval`, unless it is about fundamental Mathics objects (Strings, Integer, Patterns, etc.) which would go into  `mathics.core`.

**Comment by mmatera** [comment]
Ok, in that case, keep what indicates your version.

File: mathics/builtin/numbers/algebra.py [hunk 7]
```diff
@@ -899,6 +899,8 @@ class CoefficientList(Builtin):
      = {{5, d, 0, b}, {c, 0, 0, 0}, {a, 0, 0, 0}}
     >> CoefficientList[(x - 2 y + 3 z)^3, {x, y, z}]
      = {{{0, 0, 0, 27}, {0, 0, -54, 0}, {0, 36, 0, 0}, {-8, 0, 0, 0}}, {{0, 0, 27, 0}, {0, -36, 0, 0}, {12, 0, 0, 0}, {0, 0, 0, 0}}, {{0, 9, 0, 0}, {-6, 0, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}}, {{1, 0, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}}}
+    >> CoefficientList[Series[Log[1-x], {x, 0, 9}], x]
+     = {-1, -1 / 2, -1 / 3, -1 / 4, -1 / 5, -1 / 6, -1 / 7, -1 / 8, -1 / 9}
```
Comment (axkr) [inline mathics/builtin/numbers/algebra.py:903]: this should return:
`= {`0, -1, -1 / 2, -1 / 3, -1 / 4, -1 / 5, -1 / 6, -1 / 7, -1 / 8, -1 / 9}

**Comment by rocky** [comment]
@davidar There have been a few comments here. We could split this PR into two to four, for example ordered by guide section that the functions fall in. That would make it easier to focus on smaller parts for review. 

(I like to pull down PRs and look at them inside Django to see how the docstrings render and ensure the doc examples flow naturally. So here, smaller is better.)

File: mathics/builtin/intfns/combinatorial.py [hunk 8]
```diff
@@ -39,6 +40,35 @@
 SymbolSubsets = Symbol("Subsets")
 
 
+class BellB(SympyFunction):
+    """
+    <url>:Bell number: https://en.wikipedia.org/wiki/Bell_number</url>
+    (<url>:SymPy: https://docs.sympy.org/latest/modules/functions/combinatorial.html#sympy.functions.combinatorial.numbers.bell</url>,
```
Comment (rocky) [inline mathics/builtin/intfns/combinatorial.py:46]: Remove newline before `(<url>`. and after `<url>,` 

Sadly line breaks in in the docstring are hard newlines. So this renders as 

image

File: mathics/builtin/intfns/combinatorial.py [hunk 9]
```diff
@@ -184,6 +214,35 @@ def _compute(self, n, c_ff, c_ft, c_tf, c_tt):
         )
 
 
+class EulerE(SympyFunction):
+    """
+    <url>:Euler numbers: https://en.wikipedia.org/wiki/Euler_numbers</url>
```
Comment (rocky) [inline mathics/builtin/intfns/combinatorial.py:219]: See comment on formatting for BellB.

File: mathics/builtin/intfns/combinatorial.py [hunk 10]
```diff
@@ -184,6 +214,35 @@ def _compute(self, n, c_ff, c_ft, c_tf, c_tt):
...
+    (<url>:SymPy: https://docs.sympy.org/latest/modules/functions/combinatorial.html#sympy.functions.combinatorial.numbers.euler</url>,
+    <url>:WMA: https://reference.wolfram.com/language/ref/EulerE.html</url>)
+    <dl>
+      <dt>'EulerE[$n$]'
+      <dd>Euler number $E_n$.
```
Comment (rocky) [inline mathics/builtin/intfns/combinatorial.py:224]: $E_n$ -> 'E_$n$' 
$'E_$n$($x$)

image

versus

image


versus

File: mathics/builtin/intfns/combinatorial.py (hunk 8 above)
Comment (davidar) [reply mathics/builtin/intfns/combinatorial.py:46]: Oh, I wondered why some of the other builtins had newlines in weird places, that's annoying

Comment (rocky) [reply mathics/builtin/intfns/combinatorial.py:46]: Yes, it is annoying. We gotta ditch this and replace it with a more conventional approach like Sphinx doctest. But we also have boxing issues, performance issues, and a host of other issues that just seem more important.

**Comment by rocky** [comment]
> Ok, in that case, keep what indicates your version.

Ultimately it would be ideal if we could start noting somewhere (if this doesn't already exist), the differences between versions so that someone might start a "version compatibility" package. Then packages that are written for a specific version would work.

**Comment by mmatera** [comment]
Maybe just adding a comment above the attribute definition, inside the class?

File: mathics/builtin/atomic/numbers.py (hunk 3 above)
Comment (davidar) [reply mathics/builtin/atomic/numbers.py:63]: Perhaps, in the interests of not making this PR any bigger than it already is, this could be left to another PR?

Comment (mmatera) [reply mathics/builtin/atomic/numbers.py:63]: Ok, let's do it later

**Comment by davidar** [comment]
I split out the major groups of changes into separate PRs as suggested: #1134, #1135, #1136

**Comment by rocky** [comment]
LGTM

File: mathics/core/builtin.py [hunk 11]
```diff
@@ -923,18 +926,26 @@ def eval_iter(self, expr, i, imin, imax, di, evaluation):
...
+
         result = []
-        compare_type = (
-            SymbolGreaterEqual
-            if Expression(SymbolLess, di, Integer0).evaluate(evaluation).to_python()
-            else SymbolLessEqual
-        )
         while True:
-            cont = Expression(compare_type, index, imax).evaluate(evaluation)
+            cont = Expression(SymbolLessEqual, index, normalised_range).evaluate(
```
Comment (mmatera) [inline mathics/core/builtin.py:946]: With this change, I would like to add some extra pytests checking what happens when `di=0`, or `di=Infinity`.

File: mathics/main.py [hunk 12]
```diff
@@ -124,7 +124,7 @@ def __init__(
             "NONE": (["", "", "", ""], ["", "", "", ""]),
             "LINUX": (
                 ["\033[32m", "\033[1m", "\033[0m\033[32m", "\033[39m"],
-                ["\033[31m", "\033[1m", "\033[0m\033[32m", "\033[39m"],
```
Comment (mmatera) [inline mathics/main.py:127]: Just for curiosity, why is this change here?

Comment (davidar) [reply mathics/main.py:127]: Fixes the bit in the repl where Out changes from red to green:

image

Comment (rocky) [reply mathics/main.py:127]: I see  bold red in a Gnome terminal for the default profile in Gnome Terminal: 

image

For KDE Console, I also see red: 

image


For: `xterm -bg black -fg white` here is what I see: 

image

Also, notice that what you show above is a bold orange, not a bold red. 

What terminal emulator are you running and what profile do you have?

Comment (davidar) [reply mathics/main.py:127]: I see the same issue in your screenshots: `Out[1` is red but the `]=` afterwards is green. Just a minor issue but it was annoying me ;)

Comment (rocky) [reply mathics/main.py:127]: > I see the same issue in your screenshots: `Out[1` is red but the `]=` afterwards is green. Just a minor issue but it was annoying to me ;)

Ah, okay. I see what you are talking about. Yes, you are correct; the `]=` after the red is better matched to the red that comes before.
